monix-eval/shared/src/main/scala/monix/eval/TaskLocal.scala

```diff
@@ -110,7 +110,7 @@ final class TaskLocal[A] private (default: => A) {
 
   /** Updates the local value. */
   def write(value: A): Task[Unit] =
-    Task.eval(ref.update(value))
+    Task.now(ref.update(value))
```

Comment:
Hey — so this might fix this particular issue, but we cannot do it. 

If it has eager behavior, then it is fundamentally incompatible with `Task` unfortunately.